README.md
# pymidi

A python RTP-MIDI / AppleMIDI implementation. You can use this library to build a network attached virtual MIDI device.

[![Build Status](https://travis-ci.org/mik3y/pymidi.svg?branch=main)](https://travis-ci.org/mik3y/pymidi)

**Latest release:** v0.5.0 (2020-01-12) ([changelog](https://github.com/mik3y/pymidi/blob/main/CHANGELOG.md))

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Quickstart](#quickstart)
- [Developer Setup](#developer-setup)
  - [Compatibility](#compatibility)
  - [Running tests](#running-tests)
  - [Developing against something else](#developing-against-something-else)
- [Demo Server](#demo-server)
- [Using in Another Project](#using-in-another-project)
- [Project Status](#project-status)
- [References and Reading](#references-and-reading)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Quickstart

```
$ pip install pymidi
```
or

```
poetry install pymidi
```

See [Using in Another Project](#using-in-another-project) and the [Developer Setup wiki](wiki/Developer-MIDI-Setup) for more information.

## Developer Setup

Set up your workspace with the very excellent [Poetry](https://python-poetry.org/):

```
$ poetry install
```

Once installed, you'll probably find it useful to work in a `poetry shell`, for ease of testing and running things:

```
$ poetry shell
(pymidi-tFFCbXNj)
$ python pymidi/server.py
```

### Compatibility

`pymidi` requires Python 3. It has been tested against Python 3.6 and Python 3.7.

### Running tests

Tests are run with pytest:

```
$ pytest
```

### Developing against something else

If you're working on a project that uses `pymidi` and want to develop both concurrently, leverage the setuptools `develop` command:

```
$ cd ~/git/otherproject
$ poetry shell
$ pushd ~/git/pymidi && python setup.py develop && popd
```

This creates a link to `~/git/pymidi` within the environment of `~/git/otherproject`.

## Demo Server and Examples

The library includes a simple demo server which prints stuff.

```
$ python pymidi/examples/example_server.py
```

See `--help` for usage. See the `examples/` directory for other examples.

## Using in Another Project

Most likely you will want to embed a server in another project, and respond to MIDI commands in some application specific way. The demo serve is an example of what you need to do.

First, create a subclass of `server.Handler` to implement your policy:

```py
from pymidi import server

class MyHandler(server.Handler):
    def on_peer_connected(self, peer):
        print('Peer connected: {}'.format(peer))

    def on_peer_disconnected(self, peer):
        print('Peer disconnected: {}'.format(peer))

    def on_midi_commands(self, peer, midi_packet):
        for command in midi_packet.command.midi_list:
            if command.command == 'note_on':
                key = command.params.key
                velocity = command.params.velocity
                print('Someone hit the key {} with velocity {}'.format(key, velocity))
```

Then install it in a server and start serving:

```
myServer = server.Server([('0.0.0.0', 5051)])
myServer.add_handler(MyHandler())
myServer.serve_forever()
```

See the [Developer Setup wiki](https://github.com/mik3y/pymidi/wiki/Developer-MIDI-Setup) for ways to test with real devices.

## Project Status

What works:
* Exchange packet parsing
* Timestamp sync packet parsing
* Exchange & timestamp sync protocol support
* MIDI message parsing

Not (yet) implemented:
* Journal contents parsing
* Verification of peers on the data channel
* Auto-disconnect peers that stop synchronizing clocks

## References and Reading

* Official docs
  - [RFC 6295: RTP Payload Format for MIDI](https://tools.ietf.org/html/rfc6295)
  - [AppleMIDI Reference Documentation from Apple](https://developer.apple.com/library/archive/documentation/Audio/Conceptual/MIDINetworkDriverProtocol/MIDI/MIDI.html)
  - [RTP-MIDI on Wikipedia](https://en.wikipedia.org/wiki/RTP-MIDI)
* Other helpful docs/sites
  - [The MIDI Specification](http://midi.teragonaudio.com/tech/midispec.htm)


CHANGELOG.md
# Changelog

## Current version (in development)

* Bugfix: Fix crash when sender name contains non-ascii characters (#18).
* Internal: Switched from `pipenv` to `poetry`.
* Internal: Added `black` for code formatting.
* Improvement: Calculates latency to client and outputs as info.
* Improvement: Decodes pitch blend change messages.
* Breaking change: `on_midi_commands` callback handler now passes whole MIDI packet rather than a list of commands - this is useful if the journal (or other data) is required.
* Improvement: RTP sequence numbers now increment
* Improvement: Timestamps don't overflow 32-bit field (issue #34)
* Improvement: disconnt() method for client (issue #28)

## v0.5.0 (2020-01-12)

* Python 2 support removed.

## v0.4.0 (2018-12-26)

* Improvement: Python 3 support (#9).
* Bugfix: Demo server: Fix IPv4/IPv6 support in dualstack environments (#8).

## v0.3.0 (2018-10-20)

* Improvement: Server instances can bind to ipv6 addresses.

## v0.2.1 (2018-09-16)

* Repackaged release, no functional changes.

## v0.2.0 (2018-09-16)

* `note_on` and `note_off` messages now report a string note name.
* Cleaned up some logs.

## v0.1.0 (2018-09-16)

* First release.


pymidi/tests/utils_test.py
from unittest import TestCase
from pymidi import utils


class UtilsTests(TestCase):
    def test_is_ipv4_address(self):
        bad_ips = ['hello', '', None, object]
        for ip in bad_ips:
            self.assertEqual(
                False, utils.is_ipv4_address(ip), 'Expected {} to return False'.format(repr(ip))
            )

        good_ips = ['127.0.0.1', '8.8.8.8']
        for ip in good_ips:
            self.assertEqual(
                True, utils.is_ipv4_address(ip), 'Expected {} to return True'.format(repr(ip))
            )

    def test_is_ipv6_address(self):
        bad_ips = ['hello', '', None, object]
        for ip in bad_ips:
            self.assertEqual(
                False, utils.is_ipv6_address(ip), 'Expected {} to return False'.format(repr(ip))
            )

        good_ips = ['::', '2001:0db8:85a3:0000:0000:8a2e:0370:7334']
        for ip in good_ips:
            self.assertEqual(
                True, utils.is_ipv6_address(ip), 'Expected {} to return True'.format(repr(ip))
            )

    def test_validate_addr(self):
        bad_addrs = [None, ('boom', 'dip'), ('', 80), ('localhost', 80)]
        for addr in bad_addrs:
            with self.assertRaises(ValueError):
                utils.validate_addr(addr)

        good_addrs = [('127.0.0.1', 80), ('8.8.8.8', 2048), ('::', 5051)]
        for addr in good_addrs:
            utils.validate_addr(addr)

    def test_b2h(self):
        mybytes = b'yo'
        self.assertEqual('796f', utils.b2h(mybytes))
        mybytes = b'\xfe\xed\xfa\xce'
        self.assertEqual('feedface', utils.b2h(mybytes))


pymidi/tests/packets_test.py
from unittest import TestCase
from pymidi import packets
from pymidi.utils import h2b

EXCHANGE_PACKET = h2b('ffff494e000000026633487347d810964d696b65e2809973204d616300')
TIMESTAMP_PACKET = h2b('ffff434b47d8109602000000000000004400227e00000dfaad1e5c820000000044002288')

# A midi packet with a Note On command
SINGLE_MIDI_PACKET = h2b('8061427a4b9f303647d8109643903026204276000608006685')

# A midi packet with two commands, utilizing running status.
#
# Commands
#   0x90 0x3e 0x41  - NOTE_ON D3 velocity 49
#   0x40 0x3b       - NOTE_ON E3 velocity 59
#
# Journal
#   NOTE_ON C3 velocity 37
MULTI_MIDI_PACKET = h2b(
    '8061429a51d2dc8747d8109646903e310a403b21427c00090881673c250d50c8060880440e'
)

# A controller/mode change packet
CONTROL_MODE_CHANGE_PACKET = h2b('80614ba55944067e47d8109643b06c00204ba0000948006c7f807708')

# An ApplieMIDI invitation packet
APPLEMIDI_INVITATION_PACKET = h2b('ffff494e000000020507236747d810966d626f6f6b2d73657373696f6e00')

# An AppleMIDI exit packet
APPLEMIDI_EXIT_PACKET = h2b('ffff4259000000020000000047d81096')


class TestPackets(TestCase):
    def test_exchange_packet(self):
        pkt = packets.AppleMIDIExchangePacket.parse(EXCHANGE_PACKET)
        self.assertEqual(b'\xff\xff', pkt.preamble)
        self.assertEqual(b'IN', pkt.command)
        self.assertEqual(2, pkt.protocol_version)
        self.assertEqual(1714636915, pkt.initiator_token)
        self.assertEqual(1205342358, pkt.ssrc)
        self.assertEqual('Mike’s Mac', pkt.name)

    def test_timestamp_packet(self):
        pkt = packets.AppleMIDITimestampPacket.parse(TIMESTAMP_PACKET)
        self.assertEqual(b'\xff\xff', pkt.preamble)
        self.assertEqual(b'CK', pkt.command)
        self.assertEqual(1205342358, pkt.ssrc)
        self.assertEqual(2, pkt.count)
        self.assertEqual(1140859518, pkt.timestamp_1)
        self.assertEqual(15370297433218, pkt.timestamp_2)
        self.assertEqual(1140859528, pkt.timestamp_3)

    def test_single_midi_packet(self):
        pkt = packets.MIDIPacket.parse(SINGLE_MIDI_PACKET)
        print(pkt)
        self.assertTrue(pkt.header, 'Expected a header')
        self.assertEqual(2, pkt.header.rtp_header.flags.v)
        self.assertEqual(False, pkt.header.rtp_header.flags.p)
        self.assertEqual(False, pkt.header.rtp_header.flags.x)
        self.assertEqual(0, pkt.header.rtp_header.flags.cc)
        self.assertEqual(False, pkt.header.rtp_header.flags.m)
        self.assertEqual(0x61, pkt.header.rtp_header.flags.pt)
        self.assertEqual(17018, pkt.header.rtp_header.sequence_number)

        self.assertTrue(pkt.command, 'Expected a command')
        self.assertEqual(False, pkt.command.flags.b)
        self.assertEqual(True, pkt.command.flags.j)
        self.assertEqual(False, pkt.command.flags.z)
        self.assertEqual(False, pkt.command.flags.p)
        self.assertEqual(3, pkt.command.flags.len)
        print(pkt.command)
        self.assertEqual(1, len(pkt.command.midi_list))

        command = pkt.command.midi_list[0]
        self.assertEqual(0x90, command.command_byte)
        self.assertEqual('note_on', command.command)
        self.assertEqual('C3', command.params.key)
        self.assertTrue(38, command.params.velocity)

        self.assertTrue(pkt.journal, 'Expected journal')

    def test_multi_midi_packet(self):
        pkt = packets.MIDIPacket.parse(MULTI_MIDI_PACKET)
        self.assertTrue(pkt.header, 'Expected a header')
        self.assertEqual(2, pkt.header.rtp_header.flags.v)
        self.assertEqual(False, pkt.header.rtp_header.flags.p)
        self.assertEqual(False, pkt.header.rtp_header.flags.x)
        self.assertEqual(0, pkt.header.rtp_header.flags.cc)
        self.assertEqual(False, pkt.header.rtp_header.flags.m)
        self.assertEqual(0x61, pkt.header.rtp_header.flags.pt)
        self.assertEqual(17050, pkt.header.rtp_header.sequence_number)

        self.assertTrue(pkt.command, 'Expected a command')
        self.assertEqual(False, pkt.command.flags.b)
        self.assertEqual(True, pkt.command.flags.j)
        self.assertEqual(False, pkt.command.flags.z)
        self.assertEqual(False, pkt.command.flags.p)
        self.assertEqual(6, pkt.command.flags.len)
        self.assertEqual(2, len(pkt.command.midi_list))

        command = pkt.command.midi_list[0]
        self.assertEqual(0x90, command.command_byte)
        self.assertEqual('note_on', command.command)
        self.assertEqual('D4', command.params.key)
        self.assertTrue(38, command.params.velocity)

        command = pkt.command.midi_list[1]
        self.assertEqual(0x90, command.command_byte)
        self.assertEqual('note_on', command.command)
        self.assertEqual('E4', command.params.key)
        self.assertTrue(38, command.params.velocity)

        self.assertTrue(pkt.journal, 'Expected journal')

    def test_packet_with_no_journal(self):
        pkt = packets.MIDIPacket.parse(h2b('806142a0550d8a5a47d8109603903446'))
        self.assertEqual(False, pkt.command.flags.j, 'Expected J bit to be clear')
        self.assertTrue(not pkt.journal, 'Expected no journal')

    def test_to_string(self):
        pkt = packets.MIDIPacket.parse(SINGLE_MIDI_PACKET)
        strval = packets.to_string(pkt)
        self.assertEqual('MIDIPacket [note_on C3 38]', strval)

        pkt = packets.MIDIPacket.parse(MULTI_MIDI_PACKET)
        strval = packets.to_string(pkt)
        self.assertEqual('MIDIPacket [note_on D4 49] [note_on E4 59]', strval)

        pkt = packets.MIDIPacket.parse(CONTROL_MODE_CHANGE_PACKET)
        strval = packets.to_string(pkt)
        self.assertEqual('MIDIPacket [control_mode_change 108 0]', strval)

        pkt = packets.AppleMIDIExchangePacket.parse(APPLEMIDI_INVITATION_PACKET)
        strval = packets.to_string(pkt)
        self.assertEqual(
            'AppleMIDIExchangePacket [command=IN ssrc=1205342358 name=mbook-session]', strval
        )

        pkt = packets.AppleMIDIExchangePacket.parse(APPLEMIDI_EXIT_PACKET)
        strval = packets.to_string(pkt)
        self.assertEqual('AppleMIDIExchangePacket [command=BY ssrc=1205342358 name=None]', strval)


pymidi/tests/server_tests.py
from unittest import TestCase
from pymidi.server import Server, Handler
import mock


class FakeHandler(mock.Mock, Handler):
    pass


class ServerTests(TestCase):
    def setUp(self):
        self.server = Server([('127.0.0.1', 0)])
        self.server._init_protocols()
        self.handler = FakeHandler()
        self.server.add_handler(self.handler)

    def test_server_bind(self):
        protos = self.server.socket_map
        self.assertEqual(2, len(protos))

    def test_loop_once_no_data(self):
        """Confirms a single read loop succeeds with no data."""
        self.server._loop_once(timeout=0)
        self.assertFalse(self.handler.called)
        self.assertEqual(0, self.handler.call_count)


pymidi/utils.py
import codecs
import socket
from six import string_types
from builtins import bytes


def h2b(s):
    """Converts a hex string to bytes; Python 2/3 compatible."""
    return bytes.fromhex(s)


def b2h(b):
    """Converts a `bytes` object to a hex string."""
    if not isinstance(b, bytes):
        raise ValueError('Argument must be a `bytes`')
    result = codecs.getencoder('hex_codec')(b)[0]
    if isinstance(result, bytes):
        result = result.decode('ascii')
    return result


def is_ipv4_address(ipstr):
    try:
        socket.inet_aton(ipstr)
        return True
    except (socket.error, TypeError):
        return False


def is_ipv6_address(ipstr):
    try:
        socket.inet_pton(socket.AF_INET6, ipstr)
        return True
    except (socket.error, TypeError):
        return False


def is_ipv4_or_ipv6_address(ipstr):
    return is_ipv4_address(ipstr) or is_ipv6_address(ipstr)


def validate_addr(addr):
    """Raises `ValueError` if `addr` is not a well-formed (ip, port) pair."""
    if not isinstance(addr, tuple):
        raise ValueError('Address {} is not a tuple'.format(repr(addr)))
    if len(addr) != 2:
        raise ValueError('Address {} is not a 2-tuple'.format(repr(addr)))
    if not isinstance(addr[0], string_types):
        raise ValueError('First param of address {} is not a string'.format(repr(addr)))
    if not is_ipv4_or_ipv6_address(addr[0]):
        raise ValueError('First param of address {} is not a valid ip'.format(repr(addr)))
    if not isinstance(addr[1], int):
        raise ValueError('Second param of address {} is not an int'.format(repr(addr)))


pymidi/client.py
from builtins import bytes

from optparse import OptionParser
import logging
import select
import socket
import sys
import random

from pymidi import packets
from pymidi import protocol
from pymidi import utils
from pymidi.utils import b2h
from pymidi.utils import get_timestamp
from construct import ConstructError

try:
    import coloredlogs
except ImportError:
    coloredlogs = None

logger = logging.getLogger('pymidi.client')


class ClientError(Exception):
    """General client error."""


class AlreadyConnected(ClientError):
    """Client is already connected."""


class Client(object):
    def __init__(self, name='PyMidi', ssrc=None, sourcePort=None):
        """Creates a new Client instance."""
        self.ssrc = ssrc or random.randint(0, 2 ** 32 - 1)
        self.socket = [None,None] # Need to have a command and data socket on the client side
        self.host = None
        self.port = None
        self.sourcePort = sourcePort or 5004
        self.name = name or 'PyMidi'
        self.sequenceNumber = 1

    def connect(self, host, port):
        if self.host and self.port:
            raise ClientError(f'Already connected to {self.host}:{self.port}')

        pkt = packets.AppleMIDIExchangePacket.create(
            protocol_version=2,
            command=protocol.APPLEMIDI_COMMAND_INVITATION,
            initiator_token=random.randint(0, 2 ** 32 - 1),
            ssrc=self.ssrc,
            name=self.name
        )

        for index in (0, 1):
            self.socket[index] = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.socket[index].bind(('0.0.0.0', self.sourcePort+index))

            logger.info(f'Sending exchange packet to port {port+index} from port {self.sourcePort+index}...')
            self.socket[index].sendto(pkt, (host, port+index))
            packet = self.get_next_packet(self.socket[index])
            if not packet:
                raise Exception('No packet received')
            if packet._name != 'AppleMIDIExchangePacket':
                raise Exception('Expected exchange packet')
            logger.info(f'Exchange successful.')

        self.host = host
        self.port = port

    def disconnect(self):
        if not self.socket[0]:
            raise ClientError(f'Not connected to anywhere')

        pkt = packets.AppleMIDIExchangePacket.create(
            protocol_version=2,
            command=protocol.APPLEMIDI_COMMAND_EXIT,
            initiator_token=0,
            ssrc=self.ssrc,
            name=None
        )
        self.socket[0].sendto(pkt, (self.host, self.port))

        for index in (0, 1): self.socket[index].close()

        self.socket = [None,None]
        self.host = None
        self.port = None

    def sync_timestamps(self, port):
        packet = packets.AppleMIDITimestampPacket.create(
            command=protocol.APPLEMIDI_COMMAND_TIMESTAMP_SYNC,
            ssrc=self.ssrc,
            count=count,
            padding=0,
            timestamp_1=get_timestamp(),
            timestamp_2=0,
            timestamp_3=0,
        )

    def send_note_on(self, notestr, velocity=80, channel=1):
        self._send_note(notestr, packets.COMMAND_NOTE_ON, velocity, channel)

    def send_note_off(self, notestr, velocity=80, channel=1):
        self._send_note(notestr, packets.COMMAND_NOTE_OFF, velocity, channel)

    def _send_note(self, notestr, command, velocity=80, channel=1):
        # key = packets.MIDINote.build(notestr)
        command = {
            'flags': {
                'b': 0,
                'j': 0,
                'z': 0,
                'p': 0,
                'len': 3,
            },
            'midi_list': [
                {
                    'delta_time': 0,
                    '__next': 0x80,  # TODO(mikey): This shouldn't be needed.
                    'command': 'note_on' if command == packets.COMMAND_NOTE_ON else 'note_off',
                    'command_byte': command | (channel & 0xF),
                    'channel': channel,
                    'params': {
                        'key': notestr,
                        'velocity': velocity,
                    },
                }
            ],
        }
        self._send_rtp_command(self.socket[1], command)

    def _send_rtp_command(self, socket, command):
        packet = packets.MIDIPacket.create(
            header={
                'rtp_header': {
                    'flags': {
                        'v': 0x2,
                        'p': 0,
                        'x': 0,
                        'cc': 0,
                        'm': 0x1,
                        'pt': 0x61,
                    },
                    'sequence_number': self.sequenceNumber,
                },
                'timestamp': get_timestamp(),
                'ssrc': self.ssrc,
            },
            command=command,
            journal='',
        )

        socket.sendto(packet, (self.host, self.port + 1))
        self.sequenceNumber += 1

    def get_next_packet(self, socket):
        data, addr = socket.recvfrom(1024)
        command = data[2:4]
        try:
            if data[0:2] == protocol.APPLEMIDI_PREAMBLE:
                command = data[2:4]
                logger.debug('Command: {}'.format(b2h(command)))
                return self.handle_command_message(command, data, addr)
        except ConstructError:
            logger.exception('Bug or malformed packet, ignoring')
        return None

    def handle_command_message(self, command, data, addr):
        if command == protocol.APPLEMIDI_COMMAND_INVITATION_ACCEPTED:
            return packets.AppleMIDIExchangePacket.parse(data)
        else:
            logger.warning('Ignoring unrecognized command: {}'.format(command))
        return None


pymidi/server.py
from builtins import bytes

from optparse import OptionParser
import logging
import select
import socket
import sys

from pymidi.protocol import DataProtocol
from pymidi.protocol import ControlProtocol
from pymidi import utils

try:
    import coloredlogs
except ImportError:
    coloredlogs = None

logger = logging.getLogger('pymidi.server')


class Handler(object):
    def on_peer_connected(self, peer):
        pass

    def on_peer_disconnected(self, peer):
        pass

    def on_midi_commands(self, peer, midi_packet):
        pass


class Server(object):
    def __init__(self, bind_addrs):
        """Creates a new Server instance.

        `bind_addrs` should be an iterable of 1 or more addresses to bind to,
        each a 2-tuple of (ip, port). Socket family will be automatically
        detected from the IP address.
        """
        if not bind_addrs:
            raise ValueError('Must provide at least one bind address.')
        map(utils.validate_addr, bind_addrs)
        self.bind_addrs = bind_addrs
        self.handlers = set()

        # Maps sockets to their protocol handlers.
        self.socket_map = {}

    @classmethod
    def from_bind_addrs(cls, hosts):
        """Convenience method to construct an instance from a string."""
        bind_addrs = set()
        for host in hosts:
            parts = host.split(':')
            name = ':'.join(parts[:-1])
            port = int(parts[-1])
            addr = (name, port)
            bind_addrs.add(addr)
        return cls(bind_addrs)

    def add_handler(self, handler):
        assert isinstance(handler, Handler)
        self.handlers.add(handler)

    def remove_handler(self, handler):
        assert isinstance(handler, Handler)
        self.handlers.discard(handler)

    def _peer_connected_cb(self, peer):
        for handler in self.handlers:
            handler.on_peer_connected(peer)

    def _peer_disconnected_cb(self, peer):
        for handler in self.handlers:
            handler.on_peer_disconnected(peer)

    def _midi_command_cb(self, peer, midi_packet):
        for handler in self.handlers:
            handler.on_midi_commands(peer, midi_packet)

    def _build_control_protocol(self, host, port, family):
        logger.info('Control socket on {}:{}'.format(host, port))
        control_socket = socket.socket(family, socket.SOCK_DGRAM)
        control_socket.bind((host, port))
        return ControlProtocol(
            socket=control_socket,
            connect_cb=self._peer_connected_cb,
            disconnect_cb=self._peer_disconnected_cb,
        )

    def _build_data_protocol(self, host, family, ctrl_protocol):
        ctrl_port = ctrl_protocol.socket.getsockname()[1]
        logger.info('Data socket on {}:{}'.format(host, ctrl_port + 1))
        data_socket = socket.socket(family, socket.SOCK_DGRAM)
        data_socket.bind((host, ctrl_port + 1))
        data_protocol = DataProtocol(data_socket, midi_command_cb=self._midi_command_cb)
        ctrl_protocol.associate_data_protocol(data_protocol)
        return data_protocol

    def _init_protocols(self):
        for host, port in self.bind_addrs:
            if utils.is_ipv4_address(host):
                family = socket.AF_INET
            elif utils.is_ipv6_address(host):
                family = socket.AF_INET6
            else:
                raise ValueError('Invalid bind host: "{}"'.format(host))

            ctrl_protocol = self._build_control_protocol(host, port, family)
            data_protocol = self._build_data_protocol(host, family, ctrl_protocol)

            self.socket_map[data_protocol.socket] = data_protocol
            self.socket_map[ctrl_protocol.socket] = ctrl_protocol

            protos = (ctrl_protocol, data_protocol)
            if family == socket.AF_INET:
                self.ipv4_protocols = protos
            elif family == socket.AF_INET6:
                self.ipv6_protocols = protos

    def _loop_once(self, timeout=None):
        sockets = self.socket_map.keys()
        rr, _, _ = select.select(sockets, [], [], timeout)
        for s in rr:
            buffer, addr = s.recvfrom(1024)
            buffer = bytes(buffer)
            proto = self.socket_map[s]
            proto.handle_message(buffer, addr)

    def serve_forever(self):
        self._init_protocols()
        while True:
            self._loop_once()


pymidi/packets.py
from construct import Struct as BaseStruct
from construct import Const, CString, Padding, Int8ub, Int16ub, Int32ub
from construct import Int64ub, Bitwise, BitStruct, BitsInteger, Nibble, Flag, Optional, Bytes
from construct import If, IfThenElse, GreedyBytes, GreedyRange, VarInt, FixedSized, Byte, Computed
from construct import Switch, Enum, Peek
from construct import this as _this


COMMAND_NOTE_OFF = 0x80
COMMAND_NOTE_ON = 0x90
COMMAND_AFTERTOUCH = 0xA0
COMMAND_CONTROL_MODE_CHANGE = 0xB0
COMMAND_PITCH_BEND_CHANGE = 0xE0


def to_string(pkt):
    """Pretty-prints a packet."""
    name = pkt._name
    detail = ''

    if name == 'AppleMIDIExchangePacket':
        detail = '[command={} ssrc={} name={}]'.format(
            pkt.command.decode('utf-8'), pkt.ssrc, pkt.name
        )
    elif name == 'MIDIPacket':
        items = []
        for entry in pkt.command.midi_list:
            command = entry.command
            if command in ('note_on', 'note_off'):
                items.append('{} {} {}'.format(command, entry.params.key, entry.params.velocity))
            elif command == 'control_mode_change':
                items.append(
                    '{} {} {}'.format(command, entry.params.controller, entry.params.value)
                )
            elif command == 'aftertouch':
                items.append('{} {} {}'.format(command, entry.params.key, entry.params.touch))
            elif command == 'pitch_bend_change':
                items.append('{} {} {}'.format(command, entry.params.lsb, entry.params.msb))
            else:
                items.append(command)
        detail = ' '.join(('[{}]'.format(i) for i in items))

    return '{} {}'.format(name, detail)


def remember_last(obj, ctx):
    """Stores the last-seen command byte in the parsing context.

    Bit of a hack to make running status support work.
    """
    setattr(ctx._root, '_last_command_byte', obj)


class Struct(BaseStruct):
    """Adds `create()`, a friendlier `build()` method."""

    def create(self, **kwargs):
        return self.build(kwargs)


AppleMIDIExchangePacket = Struct(
    '_name' / Computed('AppleMIDIExchangePacket'),
    'preamble' / Const(b'\xff\xff'),
    'command' / Bytes(2),
    'protocol_version' / Int32ub,
    'initiator_token' / Int32ub,
    'ssrc' / Int32ub,
    'name' / Optional(CString('utf8')),
)

AppleMIDITimestampPacket = Struct(
    '_name' / Computed('AppleMIDITimestampPacket'),
    'preamble' / Const(b'\xff\xff'),
    'command' / Bytes(2),
    'ssrc' / Int32ub,
    'count' / Int8ub,
    'padding' / Padding(3),
    'timestamp_1' / Int64ub,
    'timestamp_2' / Int64ub,
    'timestamp_3' / Int64ub,
)

MIDIPacketHeaderFlags = Bitwise(
    Struct(
        'v' / BitsInteger(2),  # always 0x2
        'p' / Flag,  # always 0
        'x' / Flag,  # always 0
        'cc' / Nibble,  # always 0
        'm' / Flag,  # always 0x1
        'pt' / BitsInteger(7),  # always 0x61
    )
)

RTPHeader = Struct(
    'flags' / MIDIPacketHeaderFlags,
    'sequence_number' / Int16ub,
)

MIDIPacketHeader = Struct(
    '_name' / Computed('MIDIPacketHeader'),
    'rtp_header' / RTPHeader,
    'timestamp' / Int32ub,
    'ssrc' / Int32ub,
)

MIDINote = Enum(
    Byte,
    Cn1=0,
    Csn1=1,
    Dn1=2,
    Dsn1=3,
    En1=4,
    Fn1=5,
    Fsn1=6,
    Gn1=7,
    Gsn1=8,
    An1=9,
    Asn1=10,
    Bn1=11,
    C0=12,
    Cs0=13,
    D0=14,
    Ds0=15,
    E0=16,
    F0=17,
    Fs0=18,
    G0=19,
    Gs0=20,
    A0=21,
    As0=22,
    B0=23,
    C1=24,
    Cs1=25,
    D1=26,
    Ds1=27,
    E1=28,
    F1=29,
    Fs1=30,
    G1=31,
    Gs1=32,
    A1=33,
    As1=34,
    B1=35,
    C2=36,
    Cs2=37,
    D2=38,
    Ds2=39,
    E2=40,
    F2=41,
    Fs2=42,
    G2=43,
    Gs2=44,
    A2=45,
    As2=46,
    B2=47,
    C3=48,
    Cs3=49,
    D3=50,
    Ds3=51,
    E3=52,
    F3=53,
    Fs3=54,
    G3=55,
    Gs3=56,
    A3=57,
    As3=58,
    B3=59,
    C4=60,
    Cs4=61,
    D4=62,
    Ds4=63,
    E4=64,
    F4=65,
    Fs4=66,
    G4=67,
    Gs4=68,
    A4=69,
    As4=70,
    B4=71,
    C5=72,
    Cs5=73,
    D5=74,
    Ds5=75,
    E5=76,
    F5=77,
    Fs5=78,
    G5=79,
    Gs5=80,
    A5=81,
    As5=82,
    B5=83,
    C6=84,
    Cs6=85,
    D6=86,
    Ds6=87,
    E6=88,
    F6=89,
    Fs6=90,
    G6=91,
    Gs6=92,
    A6=93,
    As6=94,
    B6=95,
    C7=96,
    Cs7=97,
    D7=98,
    Ds7=99,
    E7=100,
    F7=101,
    Fs7=102,
    G7=103,
    Gs7=104,
    A7=105,
    As7=106,
    B7=107,
    C8=108,
    Cs8=109,
    D8=110,
    Ds8=111,
    E8=112,
    F8=113,
    Fs8=114,
    G8=115,
    Gs8=116,
    A8=117,
    As8=118,
    B8=119,
    C9=120,
    Cs9=121,
    D9=122,
    Ds9=123,
    E9=124,
    F9=125,
    Fs9=126,
    G9=127,
)

MIDIPacketCommand = Struct(
    '_name' / Computed('MIDIPacketCommand'),
    'flags'
    / BitStruct(
        'b' / Flag,
        'j' / Flag,
        'z' / Flag,
        'p' / Flag,
        'len' / IfThenElse(_this.b == 0, BitsInteger(4), BitsInteger(12)),
    ),
    # 'midi_list' / Bytes(_this.flags.len),
    'midi_list'
    / FixedSized(
        _this.flags.len,
        GreedyRange(
            Struct(
                'delta_time' / If(_this._index > 0, VarInt),
                # The "running status" technique means multiple commands may be sent under
                # the same status. This condition occurs when, after parsing the current
                # commands, we see the next byte is NOT a status byte (MSB is low).
                #
                # Below, this is accomplished by storing the most recent status byte
                # on the global context with the `* remember_last` macro; then using it
                # on the `else` branch of the `command_byte` selection.
                '__next' / Peek(Int8ub),
                'command_byte'
                / IfThenElse(
                    _this.__next & 0x80,
                    Byte * remember_last,
                    Computed(lambda ctx: ctx._root._last_command_byte),
                ),
                'command'
                / If(
                    _this.command_byte,
                    Enum(
                        Computed(_this.command_byte & 0xF0),
                        note_on=COMMAND_NOTE_ON,
                        note_off=COMMAND_NOTE_OFF,
                        aftertouch=COMMAND_AFTERTOUCH,
                        control_mode_change=COMMAND_CONTROL_MODE_CHANGE,
                        pitch_bend_change=COMMAND_PITCH_BEND_CHANGE
                    ),
                ),
                'channel' / If(_this.command_byte, Computed(_this.command_byte & 0x0F)),
                'params'
                / Switch(
                    _this.command,
                    {
                        'note_on': Struct(
                            'key' / MIDINote,
                            'velocity' / Int8ub,
                        ),
                        'note_off': Struct(
                            'key' / MIDINote,
                            'velocity' / Int8ub,
                        ),
                        'aftertouch': Struct(
                            'key' / MIDINote,
                            'touch' / Int8ub,
                        ),
                        'control_mode_change': Struct(
                            'controller' / Int8ub,
                            'value' / Int8ub,
                        ),
                        'pitch_bend_change': Struct(
                            'lsb' / Int8ub,
                            'msb' / Int8ub,
                        ),
                    },
                    default=Struct(
                        'unknown' / GreedyBytes,
                    ),
                ),
            ),
        ),
    ),
)

MIDISystemJournal = Struct(
    '_name' / Computed('MIDISystemJournal'),
    'header'
    / BitStruct(
        's' / Flag,
        'd' / Flag,
        'v' / Flag,
        'q' / Flag,
        'f' / Flag,
        'x' / Flag,
        'length' / BitsInteger(10),
    ),
    # Note from RFC 6295 appendix A1: The "length" field includes
    # the header bytes.
    'journal' / Bytes(_this.header.length - 2),
)

MIDIChapterJournal = Struct(
    '_name' / Computed('MIDIChapterJournal'),
    'header'
    / BitStruct(
        's' / Flag,
        'chan' / BitsInteger(4),
        'h' / Flag,
        'length' / BitsInteger(10),
        'p' / Flag,
        'c' / Flag,
        'm' / Flag,
        'w' / Flag,
        'n' / Flag,
        'e' / Flag,
        't' / Flag,
        'a' / Flag,
    ),
    # Note from RFC 6295 appendix A1: The "length" field includes
    # the header bytes.
    'journal' / Bytes(_this.header.length - 3),
)

MIDIPacketJournal = Struct(
    '_name' / Computed('MIDIPacketJournal'),
    'header'
    / BitStruct(
        's' / Flag,
        'y' / Flag,
        'a' / Flag,
        'h' / Flag,
        'totchan' / BitsInteger(4),
    ),
    'checkpoint_seqnum' / Int16ub,
    'system_journal' / If(_this.header.s, MIDISystemJournal),
    'channel_journal' / If(_this.header.a, MIDIChapterJournal),
)

MIDIPacket = Struct(
    '_name' / Computed('MIDIPacket'),
    'header' / MIDIPacketHeader,
    'command' / MIDIPacketCommand,
    'journal' / If(_this.command.flags.j, MIDIPacketJournal),
)
